pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go

```diff
@@ -0,0 +1,144 @@
+package podsecuritypolicy_test
+
+import (
+	"fmt"
+	"testing"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/controllers/managementuser/rbac/podsecuritypolicy"
+	"github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3/fakes"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/version"
+)
+
+func newClusterLister(kubernetesVersion string) *fakes.ClusterListerMock {
+	return &fakes.ClusterListerMock{
+		GetFunc: func(namespace, name string) (*v3.Cluster, error) {
+			if name == "invalid" {
+				return nil, fmt.Errorf("invalid cluster: %s", name)
+			} else if name == "not ready" {
+				return &v3.Cluster{Status: v3.ClusterStatus{}}, nil
+			} else {
+				return &v3.Cluster{Status: v3.ClusterStatus{Version: &version.Info{GitVersion: kubernetesVersion}}}, nil
+			}
+		},
+	}
+}
+
+func TestCheckClusterVersionFailsForVersionsThatCannotBeParsed(t *testing.T) {
```

Comment:
> The error kinds are different for all three tests.
Right, but that just means you need different output from the function, not that you are testing something fundamentally different.

> Uniting them into one test is very messy, the assertion logic would be hard to write and read. 
I don't agree, I think the tests would look very similar to what they look like today, only with an additional value to indicate the error type desired.

> It's perfectly fine to split tests into small chunks of quick, readable statements about expected behavior.

I'm not sure I agree here either. Consider this scenario - suppose that I update this function to add a new dependency. Now, instead of updating one test, I need to look through 3 different tests, each with a different structure and different mocked dependencies. 

In short, this structure makes it harder to update the tests and the related code in the future.

Reply:
Yes, but this is perfectly fine. The failures will also alert you to the need to adjust the tests. And it all encourages to have fewer dependencies and simpler test setup.

Consider these tests here: https://github.com/golang/go/blob/ab5bd15941f3cea3695338756d0b8be0ef2321fb/src/bufio/bufio_test.go#L289.
The logic of each test is small and separate. Alternatively, the authors could have put all setup into a `setup` func, a field of test struct, but it'd be a bit clunky.